driver.go

```diff
@@ -81,7 +81,8 @@ func (d MySQLDriver) Open(dsn string) (driver.Conn, error) {
 		}
 	}
 
-	mc.buf = newBuffer(mc.netConn)
+	// Make the buffer get the context from mysqlConn.
+	mc.buf = newBufferContext(mc.netConn, &ContextWrapper{mc: mc})
```

Comment:
>Cancelleration is rare situation. Code simplicity (maintenance cost) is more important than how cancellation efficient.

I like [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and also think simplicity is important but from my point of view, I still think there is a short and blurry distance between simplicity and correctness in this case. The key point is that force to close any connections no matter they are normal or unusable is good here? I prefer that let normal one reuse again and let caller close the unusable one. It's more logical. Actually `netConn.SetDeadline()` and `UpdateDeadlineLoop` are not so hard to understand. I wrote the comments. But this is from my perspective. If you ask me to use `netConn.Close()` and remove `UpdateDeadlineLoop`, I could do it for you.

>But TCP lifetime can be shorter than mysqlConn's lifetime.

Technically, yes, but `mysqlConn.netConn`, a TCP connection, will be closed in `mysqlConn.Close()` and `mysqlConn.Close()` will be called by [`DB.putConn()`](https://github.com/golang/go/blob/release-branch.go1.8/src/database/sql/sql.go#L608) if `Pinger.Ping()` returns ErrBadConn. In other words, **Golang `database/sql/sql.go` has already provided the mechanism** to close a `driver.Conn`(`mysql.mysqlConn`), and indirectly destroy its underlying resource. Why should we do the same job again at each corner and duplicate the codes and the logic? Are there any bugs about compatibility from the earlier versions of Golang?